### ISSUE TITLE
Writing color user data doesn't work in OptiX mode

### ISSUE THREAD

**Issue by DeclanRussell** [AcademySoftwareFoundation/OpenShadingLanguage#1027]
### Problem

The way the OptiX run time exposes user data is by directly implementing the `osl_bind_interpolated_param()` function. There is already an example implementation of the callback in testrender for strings. However, it seems that if you try to use the callback for retrieving colors (and maybe other data types) the result does not get correctly passed along.

### Steps to Reproduce

1. [Here](https://github.com/imageworks/OpenShadingLanguage/files/3241176/WriteUserDataColorOptix.txt) is a patch that hard codes the user data callback so that if color user data is requested, it returns white. The patch also modifies the testoptix test so that you can simply reproduce the results with the command: testrender.exe -optix -res 320 240 scene.xml out.exr. Notice that the checkerboard.osl shader which is used for the background of the test now should simply return a color provided by user data.
2. Notice that by running the test the background is black, when the expected result should be white.
![optix_out](https://user-images.githubusercontent.com/9109767/58704862-55923b00-83a5-11e9-8581-a91f11f4e132.png)

3. Interestingly, if the user data callback returns zero, the default value of the parameter is correctly retrieved.
![ret_zero](https://user-images.githubusercontent.com/9109767/58704928-870b0680-83a5-11e9-8c92-47defc784ab9.png)

### Versions

* OSL branch/version: 1.11.0 -> master
* OS: windows
* C++ compiler: MSVC++ 14.0
* LLVM version: 8
* OIIO version: 2.1.0

**Comment by timgrant** [comment]
I'd have to dig a bit more to figure out exactly why the userdata lookup is failing, but in the meantime here is a patch that should unblock you: [userdata-fix.txt](https://github.com/imageworks/OpenShadingLanguage/files/3287412/userdata-fix.txt). It makes the userdata lookup behave more like the CPU version. Lookup by name, type, etc., is possible, but requires a bit of setup.

I could maybe roll this change into #1029, or if we encounter more userdata wrinkles we can do a new PR.

**Comment by DeclanRussell** [comment]
Thanks Tim, I just got around to trying this patch out. Looks like its solved the issue as far as I can tell, nice one!

**Comment by lgritz** [comment]
Long since solved, but this issue didn't seem to auto-close as intended.
